src/OpenApi/src/OpenApiGenerator.cs

```diff
@@ -0,0 +1,474 @@
+
+        if (methodInfo.DeclaringType is not null && !TypeHelper.IsCompilerGeneratedType(methodInfo.DeclaringType))
+        {
+            controllerName = methodInfo.DeclaringType.Name;
+        }
+        else
+        {
+            // If the declaring type is null or compiler-generated (e.g. lambdas),
+            // group the methods under the application name.
+            controllerName = _environment?.ApplicationName ?? string.Empty;
+        }
+
+        return tags is not null
+            ? tags.Tags.Select(tag => new OpenApiTag() { Name = tag }).ToList()
+            : new List<OpenApiTag>() { new OpenApiTag() { Name = controllerName } };
+    }
+
+    private List<OpenApiParameter> GetOpenApiParameters(MethodInfo methodInfo, EndpointMetadataCollection metadata, RoutePattern pattern, bool disableInferredBody)
+    {
+        var parameters = methodInfo.GetParameters();
+        var openApiParameters = new List<OpenApiParameter>();
+
+        foreach (var parameter in parameters)
+        {
+            var (isBodyOrFormParameter, parameterLocation) = GetOpenApiParameterLocation(parameter, pattern, disableInferredBody);
+
+            // If the parameter isn't something that would be populated in RequestBody
+            // or doesn't have a valid ParameterLocation, then it must be a service
+            // parameter that we can ignore.
+            if (!isBodyOrFormParameter && parameterLocation is null)
+            {
+                continue;
+            }
+
+            var nullabilityContext = new NullabilityInfoContext();
+            var nullability = nullabilityContext.Create(parameter);
+            var isOptional = parameter.HasDefaultValue || nullability.ReadState != NullabilityState.NotNull;
+            var openApiParameter = new OpenApiParameter()
+            {
+                Name = parameter.Name,
+                In = parameterLocation,
+                Content = GetOpenApiParameterContent(metadata),
+                Schema = new OpenApiSchema { Type = SchemaGenerator.GetOpenApiSchemaType(parameter.ParameterType) },
+                Required = !isOptional
+
+            };
+            openApiParameters.Add(openApiParameter);
+        }
+
+        return openApiParameters;
+    }
+
+    private static Dictionary<string, OpenApiMediaType> GetOpenApiParameterContent(EndpointMetadataCollection metadata)
+    {
+        var openApiParameterContent = new Dictionary<string, OpenApiMediaType>();
+        var acceptsMetadata = metadata.GetMetadata<IAcceptsMetadata>();
+        if (acceptsMetadata is not null)
+        {
+            foreach (var contentType in acceptsMetadata.ContentTypes)
+            {
+                openApiParameterContent.Add(contentType, new OpenApiMediaType());
+            }
+        }
+
+        return openApiParameterContent;
+    }
+
+    private (bool, ParameterLocation?) GetOpenApiParameterLocation(ParameterInfo parameter, RoutePattern pattern, bool disableInferredBody)
```

Comment:
Could you name the tuple items? Makes it easier to track what they are.

Reply:
Good call!